### PR TITLE
Add backtrace bcd

### PR DESCRIPTION
**Pull request by vasuraman11**

### CODE REVIEW

File: fqd.c [hunk 1]
```diff
@@ -158,15 +164,71 @@ static uint32_t get_my_ip(void) {
...
 }
+void signal_handler(int sig, siginfo_t *info, void *ucontext) {
+	bcd_fatal("This is a fatal crash");
+	raise(sig);
+	return;
+}
 int main(int argc, char **argv) {
+  struct sigaction new_action;
+  bcd_error_t error;
+	bcd_t bcd;
```
Comment (sbahra) [inline fqd.c]: Inconsistent spacing.

File: fqd.c [hunk 2]
```diff
@@ -158,15 +164,71 @@ static uint32_t get_my_ip(void) {
...
+  new_action.sa_handler = signal_handler;
+  sigemptyset(&new_action.sa_mask);
+  new_action.sa_flags = SA_SIGINFO;
   nodeid = get_my_ip();
   parse_cli(argc,argv);
+  if (backtrace)
```
Comment (sbahra) [inline fqd.c]: Use different variable name as to not conflict with glibc-provided `backtrace(3)`.

File: fqd.c [hunk 3]
```diff
@@ -158,15 +164,71 @@ static uint32_t get_my_ip(void) {
...
+
+    /* Initialize a handle to BCD. This should be called by every thread interacting with BCD. */
+    if (bcd_attach(&bcd, &error) == -1) {
+      fprintf(stderr, "Unable to initialize bcd handle. Exiting.\n");
+      exit(-1);
+    }
+
+    /* Set some kv pairs to be reported. */
+    char value[15];
+    snprintf(value, sizeof(value), "%u", worker_threads);
```
Comment (sbahra) [inline fqd.c]: work_threads is of type `int`, use `%d`.

File: fqd.c [hunk 4]
```diff
@@ -158,15 +164,71 @@ static uint32_t get_my_ip(void) {
...
+    if (bcd_kv(&bcd, "worker_threads", value, &error) == -1) {
+      fprintf(stderr, "Unable to set worker threads. Exiting.\n");
+      exit(-1);
+    }
+    snprintf(value, sizeof(value), "%hu", port);
+    if (bcd_kv(&bcd, "port", value, &error) == -1) {
+      fprintf(stderr, "Unable to set port. Exiting.\n");
+      exit(-1);
+    }
+    if (bcd_kv(&bcd, "application", "fq", &error) == -1) {
```
Comment (sbahra) [inline fqd.c]: This isn't necessary as ptrace will extract the process name for you.

File: fqd.c [hunk 5]
```diff
@@ -158,15 +164,71 @@ static uint32_t get_my_ip(void) {
...
+void signal_handler(int sig, siginfo_t *info, void *ucontext) {
+	bcd_fatal("This is a fatal crash");
+	raise(sig);
+	return;
+}
 int main(int argc, char **argv) {
+  struct sigaction new_action;
+  bcd_error_t error;
+	bcd_t bcd;
+  new_action.sa_handler = signal_handler;
```
Comment (sbahra) [inline fqd.c]: The signal_handler function needs to be updated to call bcd_fatal. See the README in bcd for how to call bcd_fatal.

**Comment by esproul** [comment]
Closing this in favor of more recent work.

**Pull request closed without merging**